File: cpp/algorithms/longest_common_substring/longest_common_substring.cpp
#include <algorithm>
#include <fstream>
#include <vector>


// Implementation of the longest common substring dynamic programming 
// algorithm.
// 
// The longest common substring problem is to find the longest string that is a
// substring of two or more strings.
// 
// Example: the longest common substring of the strings "ABABC" and "BABCA" is
// "ABC".
// 
// Time complexity: O(N * M), where N and M are the lengths of the two strings.
template <class Iterable>
Iterable longest_common_substring(const Iterable& A,
                                  const Iterable& B)
{
    // best[i][j] - length of the longest common substring between
    //      A[0..i] and B[0..j]
    // 
    // best[i][j] is computed as follows:
    //      best[i][j] = best[i - 1][j - 1] + 1, if A[i - 1] == B[j - 1]
    //          (we can extend the longest substring by 1)
    //      best[i][j] = max(best[i - 1][j], best[i][j - 1]) otherwise
    // 
    std::vector<std::vector<int> > best(A.size() + 1, 
            std::vector<int>(B.size() + 1));

    for (int i = 1; i <= A.size(); ++i)
        for (int j = 1; j <= B.size(); ++j) {
            if (A[i - 1] == B[j - 1]) {
                best[i][j] = best[i - 1][j - 1] + 1;
            } else {
                best[i][j] = std::max(best[i - 1][j], best[i][j - 1]);
            }
        }

    Iterable result;

    int i = A.size(), j = B.size();
    while (i != 0 && j != 0) {
        if (A[i - 1] == B[j - 1]) {
            result.push_back(A[i - 1]);
            --i, --j;
        } else if (best[i - 1][j] > best[i][j - 1]) {
            --i;
        } else {
            --j;
        }
    }

    std::reverse(result.begin(), result.end());

    return result;
}

int main()
{
    std::ifstream fin("cmlsc.in");
    std::ofstream fout("cmlsc.out");

    int N, M;
    fin >> N >> M;

    std::vector<int> A(N);
    for (int i = 0; i < N; ++i)
        fin >> A[i];

    std::vector<int> B(M);
    for (int i = 0; i < M; ++i)
        fin >> B[i];

    std::vector<int> result = longest_common_substring(A, B);

    typedef std::vector<int>::iterator iter;
    fout << result.size() << "\n";
    for (iter it = result.begin(); it != result.end(); ++it)
        fout << *it << " ";
    fout << "\n";

    return 0;
}

File: java/src/algorithms/longest_common_substring/LongestCommonSubstring.java
package algorithms.longest_common_substring;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.StringTokenizer;

/**
 * Implementation of the longest common substring dynamic programming algorithm.
 *
 * The longest common substring problem is to find the longest string that is a
 * substring of two or more strings.
 *
 * Example: the longest common substring of the strings "ABABC" and "BABCA" is
 * "ABC".
 *
 * Time complexity: O(N * M), where N and M are the lengths of the two strings.
 *
 * @author Sebastian Claici
 */
public class LongestCommonSubstring {

    public static int[] longestCommonSubstring(int[] A, int[] B) {
        /*
         * best[i][j] - length of the longest common substring between
         *      A[0..i] and B[0..j]
         *
         * best[i][j] is computed as follows:
         *      best[i][j] = best[i - 1][j - 1] + 1, if A[i - 1] == B[j - 1]
         *          (we can extend the longest substring by 1)
         *      best[i][j] = max(best[i - 1][j], best[i][j - 1]) otherwise
         */
        int[][] best = new int[A.length + 1][B.length + 1];

        for (int i = 1; i <= A.length; ++i) {
            for (int j = 1; j <= B.length; ++j) {
                if (A[i - 1] == B[j - 1]) {
                    best[i][j] = best[i - 1][j - 1] + 1;
                } else {
                    best[i][j] = Math.max(best[i - 1][j], best[i][j - 1]);
                }
            }
        }

        int size = best[A.length][B.length];
        int[] result = new int[size];

        // Build the result by walking the best array backwards
        int i = A.length, j = B.length;
        while (i != 0 && j != 0) {
            if (A[i - 1] == B[j - 1]) {
                result[--size] = A[i - 1];
                --i;
                --j;
            } else if (best[i - 1][j] > best[i][j - 1]) {
                --i;
            } else {
                --j;
            }
        }

        return result;
    }

    // Test code
    public static void main(String[] args) throws Exception {
        BufferedReader buf = new BufferedReader(new FileReader("lcs.in"));
        StringTokenizer st = new StringTokenizer(buf.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(buf.readLine());
        int[] A = new int[n];
        for (int i = 0; i < n; ++i)
            A[i] = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(buf.readLine());
        int[] B = new int[m];
        for (int i = 0; i < m; ++i)
            B[i] = Integer.parseInt(st.nextToken());

        int[] result = LongestCommonSubstring.longestCommonSubstring(A, B);
        System.out.println(result.length);
        for (int x : result)
            System.out.printf("%d ", x);
        System.out.println();
    }
}
